### ISSUE TITLE
SIGHUP should reload SSL certificates

### ISSUE THREAD

**Issue by nam-shub** [inspircd/inspircd#1143]
This was previously reported as #737 and closed with a comment that this could be done using the oper command "/rehash -ssl".

Sorry for re-opening an already closed bug, but that approach is quite limiting. Let's Encrypt is now giving away free SSL certificates with 3 month lifetimes, and encouraging automated renewal. I (and I assume many others) have a cronjob that renews the cert and reloads affected services if it's successful. For InspIRCd, my options are to have the cronjob spin up an IRC client to send OPER and REHASH, or to restart the whole ircd and disconnect users. Neither option is ideal.

**Comment by attilamolnar** [comment]
Let's Encrypt's auto renewal is fully compatible with InspIRCd, see https://github.com/inspircd/inspircd-extras/blob/master/2.0/m_rehashsslsignal.cpp which reloads ssl credentials on SIGUSR1.

I don't think changing the behavior of /REHASH to reload all certificates is a good idea because in some setups the certificates are no longer reachable after the server boots.

**Comment by attilamolnar** [comment]
Reason is that right now the reloading of the certificates and private keys are coupled together. It would be possible to allow reloading certs on `/REHASH` and the private keys on `/REHASH ssl`, but that sounds complicated.

**Comment by nam-shub** [comment]
Oh, that makes sense. Thanks for the pointer to m_rehashsslsignal, that seems to be exactly what I need.

**Comment by Krenair** [comment]
@Azelphur-Servers ran into this issue, but we're using an older (2.0.17 from mid-2014, with more recent security patches included) version of InspIRCd from the Debian package repositories. So my solution for the moment is something like this:
```
# Force IRC server rehash - password must match that at the bottom of /etc/inspircd/opers.conf
exec 3<>/dev/tcp/localhost/6667
echo -e "NICK LE" >&3
echo -e "USER LE 0 0 :LE" >&3
timeout 2s cat <&3
echo -e "OPER user secret" >&3
timeout 2s cat <&3
echo -e "REHASH -ssl" >&3
echo -e QUIT >&3
timeout 2s cat <&3
```

**Comment by SaberUK** [comment]
@Krenair Upgrade to 2.0.23 (binary packages can't build modules and we don't support the Debian packages anyway as they are a hack job) and install m_rehashsslsignal.

**Comment by Krenair** [comment]
While I sympathise with your position on the official packages entirely, we're trying to move away from software coming from places other than apt-get. Do you provide an apt repository?

**Comment by SaberUK** [comment]
@Krenair We provide experimental official packages on [the releases page](https://github.com/inspircd/inspircd/releases). They can't build extras modules yet though.

**Comment by insensitiveclod** [comment]
I'd like to subscribe to the same argument as Krenair.

The amount of different ways to build software out there and to be notified of (security) patches is increasing every year. The good thing about distributions is that they take the pain out of this process and provide you with upgrades/updates/patches in a timely fashion so that you can stay secure *and* so that you dont need to become an expert in every build-chain used out there just to maintain the installed software.

Debian has one of the better track-records and infrastructure  to aid in making this a painless process; running software outside of what's provided through Debian/other-distro is something I personally greatly refrain doing and in the companies I've worked for/with is a discouraged activity.

The LE+inspircd-ssl issue is one that's been biting me a few times now (ever 2.5 months or so..); Thank you, Krenair, for the work-around suggestion. I'll be adding that to my letsencrypt.sh postinst hooks for now.

**Comment by dijit** [comment]
*bump*

I'm looking at the ports tree on FreeBSD and there's not much in the way of sanity regarding compiling the out-of-tree module mentioned above.

Is it possible to integrate that into the base? it's only 60~ lines of code including comments.

**Comment by SISheogorath** [comment]
If I recall correct it's implemented in the current alpha 3.0. But there is no possibility to to install it to 2.0 because it's on feature freeze and only allows bugfixes.

The module exists and if you can't get it by pre-compiled packages, please compile it, yourself.

**Comment by SaberUK** [comment]
You could also ask your distribution maintainer to include it.
